### DISCUSSION TITLE
Conveyance/Delivery Decree

### DISCUSSION THREAD

**Discussion by pointycurve416** [ejektaflex/Bountiful#262] [Mod Ideas]
I would like to propose a new and unique decree type based on delivering parcels or packages to and from neighboring villages. Specifically you would deliver packages from one bounty board to another bounty board with configurable min/max distance requirements. This would help encourage exploration in custom packs and would add some new variety and depth to this mod. 

Ideas to note:
-Delivery Bounty could be textured as a package rather than the normal bounty parchment
-Rewards could have a configurable distance multiplier/time remaining bonus
-Min/max distance parameters between bounty boards to increase immersion and limit abuse in custom modpacks
-Perhaps configurable limits on how many delivery bounties can be available at a time from one bounty board to another. (example: Bounty Board A cannot have more than X delivery bounties going to Bounty Board B, so will instead begin generating bounties for Bounty Board C.)
-Possessing a Delivery Bounty could increase the spawn rate of illager patrols and cause them to target you from longer distances.
-Delivery Bounty could be used with an empty map in the inventory crafting grid to produce a map with the destinations Bounty Board, assisting players who are adverse to using coordinates, or to add compatibility with mods like Map Atlases.
-Contents of the Delivery Bounty could be irrelevant or an extension of the player made bounties idea.

**Comment by ejektaflex** [comment]
This sounds neat!
I am planning to possibly add data that uniquely identifies bounty boards, which would be required in order to tell the player which board to send them to. However, I don't yet code-wise know of a good way of going about this, for a few reasons, like:
* What do you call the destination? Just give them block coordinates?
* What if there are no other boards loaded? What if no player has travelled far enough to find another one?
* What even is a village? How can the code find one? Using the `/locate` command doesn't *actually* find villages, it finds locations with villager job sites, as far as I can tell. Even then, it only finds the closest one. Finding the ones further away could be taxing on the server, or lag.

Lots to think about when it comes to figuring out how to implement stuff like this :^)

As for illager patrols, I'd be inclined to not add them, just because I want to restrict gameplay mechanics to the boards and villagers only. If you want something similar, I'm currently in brainstorming mode for a Villager Raid objective, where mobs will attack the village.

**Comment by pointycurve416** [comment]
Forgive me in advance if I've misunderstood anything. 

  "What do you call the destination? Just give them block coordinates?"

I suppose coordinates should suffice. Players traditionally share their location in the form of coordinates as apposed to something like an address, so this feels fitting. Renaming the bounty board could alter the destination name if players so choose.
	Maybe when the package is held, a compass or map could appear somewhere on the players HUD to guide them.
 
   "What if there are no other boards loaded? What if no player has travelled far enough to find another one?"

Perhaps the board shouldn't be allowed to assign a blank decree as a delivery decree unless another board has been found within the suitable parameters, and with vanilla rates, it may be difficult to _not_ find a couple villages within reasonable distance unless it an intentional modpack design.

  "I am planning to possibly add data that uniquely identifies bounty boards, which would be required in order to tell the player which board to send them to."
  "What even is a village? How can the code find one? Using the /locate command doesn't actually find villages, it finds locations with villager job sites, as far as I can tell. Even then, it only finds the closest one. Finding the ones further away could be taxing on the server, or lag."

What sort of unique identity were you thinking of? If it has to do with identifying naturally generated bounty boards from player made boards, then perhaps limiting delivery decrees and destinations to naturally generated boards would solve this?
Maybe naturally generated bounty boards could be indestructible (within reason), and players could still create their own bounty boards that still retain the original functionality, with the exclusion of delivery decrees. Placement/destruction of natural bounty boards could be configured to /OP or creative mode only, but adding/removing decrees should still apply to survival players. 

I completely forgot that rarities exist, so instead of those perhaps they could have different package weights? These could function the same as the bounty rarity, but give them their own unique flair. Light packages could resemble envelopes whilst normal and heavy packages resemble boxes and crates.
	Additionally each tier could give varying degrees of (configurable) slowness, encouraging players to acquire proper transportation. They could also be denied access to sub inventories like ender chests or shulkers but donkeys/mules should be allowed.

**Reply by ejektaflex** [reply]
I think that this meanders a little bit too far outside of the scope of Bountiful. Not that it's a bad thing by any means, but Bountiful is only a board-and-bounty system, and new features should rather be an extension of this system. While all of this sounds the same on the surface, the addition of packages, envelopes, boxes and crates suddenly takes the mod in a different direction from where I'm intending. 

Unfortunately, I'm also not sure how to implement this into the pool system - all pool objectives have a predefined worth and range to start. Adding a dynamic objective like this doesn't fit within the current pool system architecture, unless we revamped it and added predicate checks, and picking an objective would attempt to do a validity check (in this case, trying to find another bounty to set as a destination) before deciding whether to keep or remove it. The problem is, that seems pretty expensive to do, especially if we are doing it every time a bounty is created. Villages aren't inherently close - they're usually far enough away that the world hasn't even generated them yet. We don't want to generate chunks just to find a village and bounty board (I'm sure you've played on a server where the world lags when a friend is exploring.. Imagine that lag whenever the board needs to generate a new bounty).

Additionally, bounty boards already are not tied to location. Having an objective that gets tied to going to another board a few hundred blocks away isn't interesting (as much as I hate to say it!). It's just going to make people look at a different bounty instead.

In any case, I've actually been itching to do a mail mod in the future... Forestry was one of my favorite mods because of it's mail system, and I have thought about adding something similar-ish to what you are describing in some ways! 

P.S. in the next new release, there will be an option for whether boards are breakable ;^)
